static/app/components/searchQueryBuilder/useQueryBuilderState.tsx

```diff
@@ -0,0 +1,53 @@
+import {type Reducer, useCallback, useReducer} from 'react';
+
+import type {
+  ParseResultToken,
+  Token,
+  TokenResult,
+} from 'sentry/components/searchSyntax/parser';
+
+type QueryBuilderState = {
+  focus: null; // TODO(malwilley): Implement focus state
+  query: string;
+};
+
+type DeleteTokenAction = {
+  token: ParseResultToken;
+  type: 'DELETE_TOKEN';
+};
+
+export type QueryBuilderActions = DeleteTokenAction;
+
+function removeQueryToken(query: string, token: TokenResult<Token>): string {
+  return (
+    query.substring(0, token.location.start.offset) +
+    query.substring(token.location.end.offset)
+  );
+}
+
+export function useQueryBuilderState({initialQuery}: {initialQuery: string}) {
+  const initialState: QueryBuilderState = {query: initialQuery, focus: null};
+
+  const reducer: Reducer<QueryBuilderState, QueryBuilderActions> = useCallback(
+    (state, action): QueryBuilderState => {
+      switch (action.type) {
+        case 'DELETE_TOKEN':
+          return {
+            ...state,
+            query: removeQueryToken(state.query, action.token),
+            focus: null,
+          };
+        default:
+          return state;
+      }
+    },
+    []
+  );
```

Comment:
i think you can just pull this out of the function and then you don't need useCallback

Reply:
Yeah true, I was keeping it within the hook because I might need to use other things here like org data later on